# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [532]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [533]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [534]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [535]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [536]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [537]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [538]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [539]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [540]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [541]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [542]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [543]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [544]:
df['Season'].value_counts()

Season
2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: count, dtype: int64

**Q1**: How many unique seasons are there in the dataset? Use the nunique() function.

**Answer**

To find out how many unique seasons there are in the dataset, I use the df['Season'].nunique(), as told in the exercise. With df['Season'].unique() we would get an array with all the different seasons, but with df['Season'].nunique() we get the number of different seasons. We could also use the function len(df['Season'].unique()), which also would give us the number of different seasons. Therefore, I found 32 unique seasons in the dataset.

In [545]:
# Write your code here

# print number of seasons
df['Season'].nunique()

32

**Q2**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

**Answer**

Here I find the teams with the most wins by the function df['Wteams'].value_counts(). This function prints a column in sorted order, where the first team is the one with the most wins. Therefore, I can see that the team with the most wins is 1181 with 819 wins.

In [546]:
# Write your code here

# print the teams and the number of times they have won
df['Wteam'].value_counts()


Wteam
1181    819
1242    804
1246    765
1314    761
1112    746
       ... 
1101     18
1446     14
1118      6
1289      6
1327      3
Name: count, Length: 364, dtype: int64

# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [547]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [548]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [549]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [550]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [551]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [552]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [553]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [554]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [555]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [556]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q3**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

**Answer**

Here we use the given information above. First, I sort the values in 'Season'. Then I use the season.iloc function where I insert the interval [100:201] (.iloc does not include 201), and then I want to have all columns in the dataframe.

I do this for the winning team and the winning score below:

In [557]:
# sort the values by season
season = df.sort_values('Season')

# add the index for only the rows from 100 to 201 (beacuse of iloc will now include 200) from the sorted values in season
df_season = season.iloc[100:201]
df_season

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
2525    1985     100   1406      68   1137      65    H      0
2526    1985     100   1410      83   1320      59    H      0
2527    1985     100   1423      58   1187      56    H      0
2528    1985     100   1429      86   1253      59    H      0
2529    1985     100   1440      50   1456      48    H      0
...      ...     ...    ...     ...    ...     ...  ...    ...
2447    1985      98   1212      63   1398      53    H      0
2448    1985      98   1216      64   1134      63    H      0
2449    1985      98   1233      77   1264      69    H      0
2450    1985      98   1249      80   1427      70    H      0
2425    1985      96   1449      72   1113      65    A      0

[101 rows x 8 columns]

In [558]:
# sort the values by winning team
wteam = df.sort_values('Wteam')

# add the index for only the rows from 100 to 201 (beacuse of iloc will now include 200) from the sorted values in winning team
df_Wteam = wteam.iloc[100:201]
df_Wteam

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
97792     2008       5   1102      71   1294      62    H      0
97721     2008       4   1102      67   1171      38    H      0
96770     2007     110   1102      67   1161      58    H      0
96658     2007     107   1102      69   1428      43    H      0
96454     2007     103   1102      60   1307      51    A      0
...        ...     ...    ...     ...    ...     ...  ...    ...
33628     1993      82   1102      71   1201      51    A      0
32738     1993      61   1102      69   1161      58    H      0
135550    2015      30   1102      59   1212      34    H      0
32610     1993      57   1102      75   1351      66    N      0
32310     1993      47   1102      96   1339      65    H      0

[101 rows x 8 columns]

In [559]:
# sort the values by winning score
wscore  = df.sort_values('Wscore')

# add the index for only the rows from 100 to 201 (beacuse of iloc will now include 200) from the sorted values in winning score
df_Wscore = wscore.iloc[100:201]
df_Wscore

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
99013     2008      33   1460      43   1139      42    H      0
126840    2013      77   1366      43   1126      40    H      0
126982    2013      80   1442      43   1236      40    A      0
15619     1989      30   1343      43   1159      33    H      0
122700    2012     103   1349      43   1374      39    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
135433    2015      26   1438      45   1353      26    N      0
72338     2002      90   1443      45   1117      42    H      0
135808    2015      36   1185      45   1276      42    A      0
129196    2013     131   1380      45   1341      44    N      0
132791    2014      92   1155      45   1210      41    H      0

[101 rows x 8 columns]

**Q4**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

**Answer**

I begin by combining the three subsets from earlier. After that, I remove any duplicate games that appear more than once. Then, I use .idxmax() to find the index of the highest winning score. Finally, I print the winning team and season associated with that index, which results in the season 1985, and the winning team is 1260 with a winning score of 105.


In [560]:
# Combine the three subsets 
combined_df = pd.concat([df_season, df_Wteam, df_Wscore])

# Remove duplicate rows
combined_df = combined_df.drop_duplicates()

# Find the row with the highest winning score
highest_score_row = combined_df.loc[combined_df['Wscore'].idxmax()]

# Print the results
print(f"The game with the highest winning score occurred in Season {highest_score_row['Season']}, and the winning team was {highest_score_row['Wteam']}.")
print(f"The highest winning score was {highest_score_row['Wscore']}.")


The game with the highest winning score occurred in Season 1985, and the winning team was 1260.
The highest winning score was 105.


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [561]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [562]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q5**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

**Answer**

I start by creating a new column called 'ScoreDifference'. Here I just find the absolute value of Wscore - Lscore. Then I find the average score by using the .mean() function. After that, I use the method above where I create a filtered dataframe that only includes the rows where the score difference is above the average.


In [563]:
# Write your code here

# new coloum which is the absoloute difference between winning and losing score
df['ScoreDifference'] = abs(df['Wscore'] - df['Lscore'])

# find the mean of the new coloumn
average_score = df['ScoreDifference'].mean()

# filter the dataframe so only the games with a score difference greater than 
# the average is included
filter = df[(df['ScoreDifference'] > average_score)]
filter


Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
3         1985      25   1165      70   1432      54    H      0   
6         1985      25   1228      64   1226      44    N      0   
8         1985      25   1260      98   1133      80    H      0   
10        1985      25   1307     103   1288      71    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145280    2016     131   1401      71   1261      38    N      0   
145282    2016     131   1433      76   1172      54    N      0   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference  
0                    17  
3                    16  
6                    20  
8                    18  
10                   32  
...                 ...  
145280               33  
145282               22  
145284               20  
145285               14  
145288               13  

[57227 rows x 9 columns]

**Q6**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

**Answer**

Here I use the same approach as earlier in this project, but I use the filtered dataframe.

In [564]:
# Write your code here

# find row with the heighest score difference from the dataframe filter
# from this row with the heigest score difference print the season, losing team and the winning team 
filter.iloc[filter['ScoreDifference'].argmax()] [['Season', 'Wteam','Lteam']]

Season    1996
Wteam     1409
Lteam     1341
Name: 44653, dtype: object

# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [565]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [566]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: count, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [567]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [568]:
df.values[0][0]

1985

**Q7**: Group the DataFrame by season and find the average winning score for each season.

**Answer**

Here is used the method in the description, therefore I get this:

In [569]:
# Write your code here

# i group the data by season and then find the avereage winning score for each season
df.groupby('Season')['Wscore'].mean()

Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
1990    80.846477
1991    82.733932
1992    79.992004
1993    79.508790
1994    81.160345
1995    80.096149
1996    78.205483
1997    76.300842
1998    77.229902
1999    76.043581
2000    76.310467
2001    77.088650
2002    76.963557
2003    75.795061
2004    74.971122
2005    74.693690
2006    74.587555
2007    74.829467
2008    74.962425
2009    74.033721
2010    74.652480
2011    74.374190
2012    73.379973
2013    72.840789
2014    76.101268
2015    72.680799
2016    78.295958
Name: Wscore, dtype: float64

**Q8**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

**Answer**

Instead of using the mean as before, I now use max(), then I group the data by the winning team.

In [570]:
# Write your code here

# group the data by winning team and print the maximum winning score for each team
df.groupby('Wteam')['Wscore'].max()

Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
       ... 
1460    136
1461    112
1462    125
1463    105
1464    115
Name: Wscore, Length: 364, dtype: int64

**Q9**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

**Answer**

I start out by grouping the dataframe by season and winning team, and find the mean of each winning score. Then, from this group, I find the team with the highest average winning score for each season, using .groupby(['Season']).idxmax(). With idxmax(), we find the team with the highest average winning score for each season, whereas with max(), we would get the value of the highest winning score.

In [ ]:
# Write your code here

# group the data by season and winning team
# find the average winning score for each team and season
# group this by season and find the maximum average winning score for each season 
(df.groupby(['Season','Wteam'])['Wscore'].mean()).groupby(['Season']).idxmax()

# this prints season: (season,winning team)

Season
1985    (1985, 1328)
1986    (1986, 1109)
1987    (1987, 1380)
1988    (1988, 1258)
1989    (1989, 1258)
1990    (1990, 1258)
1991    (1991, 1380)
1992    (1992, 1380)
1993    (1993, 1380)
1994    (1994, 1380)
1995    (1995, 1206)
1996    (1996, 1206)
1997    (1997, 1254)
1998    (1998, 1395)
1999    (1999, 1317)
2000    (2000, 1395)
2001    (2001, 1395)
2002    (2002, 1242)
2003    (2003, 1395)
2004    (2004, 1194)
2005    (2005, 1311)
2006    (2006, 1190)
2007    (2007, 1440)
2008    (2008, 1440)
2009    (2009, 1377)
2010    (2010, 1440)
2011    (2011, 1449)
2012    (2012, 1370)
2013    (2013, 1377)
2014    (2014, 1322)
2015    (2015, 1322)
2016    (2016, 1146)
Name: Wscore, dtype: object

**Q10**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

**Answer**

I group the data as before, but use the count() function, so it counts the number of wins in each season. Then, I convert this group to a dataframe, and for ease of understanding, I rename the column from Wscore to Wcount. Now, a dataframe is created that shows the winning team, the season, and the winning count, which represents the number of times the team won in that season.

In [572]:
# Group by and calculate the count
winning_counts = df.groupby(['Season', 'Wteam'])['Wscore'].count()

# Convert the Series to a DataFrame
df_winning_counts = winning_counts.reset_index()

# Rename the column for clarity
df_winning_counts.rename(columns={'Wscore': 'WCount'}, inplace=True)

# Display the new DataFrame
df_winning_counts


Season  Wteam  WCount
0        1985   1102       5
1        1985   1103       9
2        1985   1104      21
3        1985   1106      10
4        1985   1108      19
...       ...    ...     ...
10167    2016   1460      20
10168    2016   1461      12
10169    2016   1462      27
10170    2016   1463      21
10171    2016   1464       9

[10172 rows x 3 columns]

**Q11**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

**Answer**

For this exercise, I create a new dataframe similar to the one in the previous exercise. I grouped the data by season and winning team, then counted the winning scores for each team in each season and created a dataframe for this. After that, I found the maximum winning scores for each season using idxmax().

This dataframe shows, for each season, the winning team and the number of times the team has won in that season.

In [573]:
# Group by season and team to calculate win counts
win_counts = df.groupby(['Season', 'Wteam'])['Wscore'].count().reset_index()

# Rename the column for clarity
win_counts.rename(columns={'Wscore': 'WinCount'}, inplace=True)

# For each season, find the team with the most wins
most_wins_per_season = win_counts.loc[win_counts.groupby('Season')['WinCount'].idxmax()]

# Reset the index
most_wins_per_season = most_wins_per_season.reset_index(drop=True)

# Display the result
most_wins_per_season


Season  Wteam  WinCount
0     1985   1385        27
1     1986   1181        32
2     1987   1424        33
3     1988   1112        31
4     1989   1328        28
5     1990   1247        29
6     1991   1116        30
7     1992   1181        28
8     1993   1231        28
9     1994   1163        27
10    1995   1116        26
11    1996   1269        31
12    1997   1242        31
13    1998   1242        33
14    1999   1181        32
15    2000   1409        29
16    2001   1181        29
17    2002   1153        30
18    2003   1166        29
19    2004   1390        29
20    2005   1228        32
21    2006   1181        30
22    2007   1242        30
23    2008   1272        33
24    2009   1272        31
25    2010   1242        32
26    2011   1242        32
27    2012   1246        32
28    2013   1211        30
29    2014   1455        33
30    2015   1246        34
31    2016   1242        29

**Q12**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

**Answer**

First, I print the average losing scores across all seasons for the different teams. Then, I do the same for the average winning score.

Next, I want to create a code that makes it easy to see if there are any teams with a higher average losing score than winning score. To do this, I write code that compares the average losing and winning scores for each index. If the losing score is higher than the winning score, the code will print the team name. If not, the code will print 'There is no losing score greater than the winning score.'

When running this code, I see that all teams have a higher average winning score than losing score.

In [574]:
# Write your code here

#Define the lossing average score
losing = df.groupby('Lteam')['Lscore'].mean()
losing

Lteam
1101    60.586207
1102    59.201507
1103    64.117347
1104    64.374317
1105    61.675373
          ...    
1460    63.082915
1461    64.661972
1462    68.216117
1463    60.829213
1464    65.187063
Name: Lscore, Length: 364, dtype: float64

In [575]:
# Define the winning average score
winning = df.groupby('Wteam')['Wscore'].mean()

found = False

# A for loop so it goes through each team
for team in losing.index:
    losing_score = losing[team]
    winning_score = winning[team]
    
    # If lossing score is higher than winningscore, print which team
    if losing_score > winning_score:
        print(f"{team} has a higher average losing score ({losing_score}) than winning score ({winning_score})")
        found = True
# If found is flase, there is no lossing score higher than winningscore 
if not found:
    print("There is no losing score bigger than winning score")


There is no losing score bigger than winning score


# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [576]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q13**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

**Answer**

I create the code below, where I go through each row, and depending on that row's winning score, the new column is updated with either 'yes' or 'no'.

In [577]:
# Define a new coloum called 'HighScoringGame' that is empty
df['HighScoringGame'] = ''

# go through each index, row
for index, row in df.iterrows():
    # If the 'Wscore' in that row is greater than 100, set 'HighScoringGame' to 'Yes' for that row
    if row['Wscore'] > 100:
        df.at[index, 'HighScoringGame'] = 'Yes'
    # Else insert 'No' for that row in 'HighSciringGame'
    else:
        df.at[index, 'HighScoringGame'] = 'No'

#print dataframe
df


Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
1         1985      25   1106      77   1354      70    H      0   
2         1985      25   1112      63   1223      56    H      0   
3         1985      25   1165      70   1432      54    H      0   
4         1985      25   1192      86   1447      74    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145286    2016     132   1246      82   1401      77    N      1   
145287    2016     132   1277      66   1345      62    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference HighScoringGame  
0                    17              No  
1                     7              No  
2                     7              No  
3                    16              No  
4                    12              No  
...                 ...             ...  
145284               20              No  
145285               14              No  
145286                5              No  
145287                4              No  
145288               13              No  

[145289 rows x 10 columns]

**Q14**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

**Answer**

I start by creating an empty dictionary where I can add each team for each row, and also count the number of times each team has lost and won. This will eventually give the total amount of games each team has played.

For each row, if a team wins, I add 1 to the dictionary for that team. If the team didn’t win, nothing is added in this first part. In the second part, I add 1 if the team lost, and if they won, nothing is added. Therefore, each win or loss contributes to the total number of games played.

In [578]:
# Write your code here

# Define a empty dictionary
games = {}

# Go through each row in the dataframe
for index, row in df.iterrows():
    # If the winning team is already in the dictionary, increment its count by 1
    # Otherwise, add the team to the dictionary with a count of 1   
    if row['Wteam'] in games:
        games[row['Wteam']] += 1
    else:
        games[row['Wteam']] = 1
    # The same for lossing team; if it is in the dictionary, increase by 1, otherwise add the team with count 1
    if row['Lteam'] in games:
        games[row['Lteam']] += 1
    else:
        games[row['Lteam']] = 1

# print the dictionary
(games)


{1228: 992,
 1328: 968,
 1106: 855,
 1354: 906,
 1112: 981,
 1223: 363,
 1165: 833,
 1432: 69,
 1192: 908,
 1447: 903,
 1218: 931,
 1337: 922,
 1226: 847,
 1242: 993,
 1268: 969,
 1260: 914,
 1133: 949,
 1305: 922,
 1424: 974,
 1307: 969,
 1288: 925,
 1344: 951,
 1438: 952,
 1374: 916,
 1411: 903,
 1412: 962,
 1397: 963,
 1417: 966,
 1225: 880,
 1116: 980,
 1368: 808,
 1120: 936,
 1391: 879,
 1135: 847,
 1306: 898,
 1143: 947,
 1388: 897,
 1153: 970,
 1184: 863,
 1159: 887,
 1171: 829,
 1216: 930,
 1173: 960,
 1134: 200,
 1177: 942,
 1296: 879,
 1193: 942,
 1265: 934,
 1196: 981,
 1416: 881,
 1206: 938,
 1137: 912,
 1210: 972,
 1149: 824,
 1211: 921,
 1102: 840,
 1234: 968,
 1114: 910,
 1332: 927,
 1243: 927,
 1317: 883,
 1257: 994,
 1231: 969,
 1277: 966,
 1145: 934,
 1278: 948,
 1453: 912,
 1286: 851,
 1186: 849,
 1301: 979,
 1144: 850,
 1325: 942,
 1384: 887,
 1326: 968,
 1248: 896,
 1287: 857,
 1339: 879,
 1334: 899,
 1365: 907,
 1375: 896,
 1126: 906,
 1403: 939,
 1152: 865,
 1423

**Q15**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

**Answer**

To answer this, I create a for loop, as we learned, where I go through each row and add the highest score difference for each season. Then, I print the season, the game, and the score difference.

In [579]:
# Write your code here

# Define a empty dictionary for the highest difference per season
diff = {}

# Go trough each row in the dataframe
for index, row in df.iterrows():
    # Calculate the score difference for the current game as winning score - losing score
    score_diff = row['Wscore'] - row['Lscore']
    # Get the season for the current game
    season = row['Season']
    
    # If the season is not already in the dictionary, add it with the current game and score difference
    if season not in diff:
        diff[season] = {'game': index, 'score_diff': score_diff}
    # If the season is already in the dictionary, check if the current scoring difference is higher 
    else:
        if score_diff > diff[season]['score_diff']:
            # If it is higher update the game and the score difference
            diff[season] = {'game': index, 'score_diff': score_diff}

# for loop to print the season, game details, and score difference
for season, data in diff.items():
    # Get the index of the game with the highest score difference for this season
    game_index = data['game']
    # corresponding row from the DataFrame
    game_row = df.iloc[game_index]
    # Print the details of the season, teams, and score difference
    print(f"Season: {season}, Game: {game_row['Wteam']} vs {game_row['Lteam']}, Score Diff: {data['score_diff']}")




Season: 1985, Game: 1361 vs 1288, Score Diff: 60
Season: 1986, Game: 1314 vs 1264, Score Diff: 84
Season: 1987, Game: 1155 vs 1118, Score Diff: 73
Season: 1988, Game: 1328 vs 1147, Score Diff: 68
Season: 1989, Game: 1242 vs 1135, Score Diff: 70
Season: 1990, Game: 1181 vs 1217, Score Diff: 76
Season: 1991, Game: 1163 vs 1148, Score Diff: 68
Season: 1992, Game: 1116 vs 1126, Score Diff: 82
Season: 1993, Game: 1328 vs 1197, Score Diff: 81
Season: 1994, Game: 1228 vs 1152, Score Diff: 69
Season: 1995, Game: 1246 vs 1404, Score Diff: 74
Season: 1996, Game: 1409 vs 1341, Score Diff: 91
Season: 1997, Game: 1278 vs 1106, Score Diff: 80
Season: 1998, Game: 1395 vs 1410, Score Diff: 66
Season: 1999, Game: 1268 vs 1317, Score Diff: 75
Season: 2000, Game: 1261 vs 1212, Score Diff: 75
Season: 2001, Game: 1196 vs 1197, Score Diff: 75
Season: 2002, Game: 1328 vs 1183, Score Diff: 59
Season: 2003, Game: 1403 vs 1311, Score Diff: 72
Season: 2004, Game: 1328 vs 1115, Score Diff: 70
Season: 2005, Game: 

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [580]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q16**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

**Answer**

I use the value_counts() function for winning teams and losing teams to create two series. Then, I add them together using the function wins + losses.

In [581]:
# Write your code here

# define the wins and the losses for each team
wins = df['Wteam'].value_counts()
losses = df['Lteam'].value_counts()

# to get the total amount of games each team has played, we add then togehter (because these are series we can use +)
total_games = wins + losses
# Print total games for each team
total_games


1101     76
1102    840
1103    910
1104    975
1105    447
       ... 
1460    827
1461    914
1462    954
1463    838
1464    856
Name: count, Length: 364, dtype: int64

**Q17**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

**Answer**

I start by creating a new column in the dataframe called 'ScoreDifference', which is calculated by subtracting the losing score from the winning score. Then, I find the index for the highest score difference for each season. From this index, I print the corresponding season, teams, and the highest score difference

In [582]:
# Write your code here

# Create a new coloumn in the dataframe, scoredifference
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

# Here i find the index if the highest score difference for each season
highest_diff_idx = df.groupby('Season')['ScoreDifference'].idxmax()

# From the index above, i use them to find the row corresponing to the row with the highest score difference
highest_diff_games = df.loc[highest_diff_idx]

# print the rows where each season shows the lossing team, winning team and the highest score difference
print(highest_diff_games[['Season', 'Wteam', 'Lteam', 'ScoreDifference']])


        Season  Wteam  Lteam  ScoreDifference
236       1985   1361   1288               60
4731      1986   1314   1264               84
8240      1987   1155   1118               73
12046     1988   1328   1147               68
16677     1989   1242   1135               70
19502     1990   1181   1217               76
25161     1991   1163   1148               68
27997     1992   1116   1126               82
33858     1993   1328   1197               81
36404     1994   1228   1152               69
39858     1995   1246   1404               74
44653     1996   1409   1341               91
49033     1997   1278   1106               80
52600     1998   1395   1410               66
57618     1999   1268   1317               75
60616     2000   1261   1212               75
65999     2001   1196   1197               75
70737     2002   1328   1183               59
74791     2003   1403   1311               72
79181     2004   1328   1115               70
83372     2005   1403   1421      

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [583]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [584]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [585]:
type(df['Wscore'])

pandas.core.series.Series

In [586]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [587]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

Here's an equivalent using iloc

In [588]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [589]:
df.isnull().sum()

Season             0
Daynum             0
Wteam              0
Wscore             0
Lteam              0
Lscore             0
Wloc               0
Numot              0
ScoreDifference    0
HighScoringGame    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y